### ISSUE TITLE
dbuf_dest() destroying mutex with owner

### ISSUE THREAD

**Issue by behlendorf** [zfsonlinux/zfs#3249]
Observed after 6 hours of low memory torture testing for #3225.  This issue looks unrelated but I wanted to file it so it could be tracked.  It's likely a _very_ unlikely long standing issue.

```
2015-04-01 20:39:59 Out of memory: Kill process 141985 (maxsbrk) score 947 or sacrifice child
2015-04-01 20:39:59 Killed process 141985, UID 0, (maxsbrk) total-vm:128761676kB, anon-rss:128749440kB, file-rss:308kB
2015-04-01 20:44:00 VERIFY3(((*(volatile typeof((&db->db_mtx)->m_owner) *)&((&db->db_mtx)->m_owner))) == ((void *)0)) failed (ffff88201cf17500 == (null))
2015-04-01 20:44:00 PANIC at dbuf.c:105:dbuf_dest()
2015-04-01 20:44:00 Showing stack for process 60530
2015-04-01 20:44:00 Pid: 60530, comm: rm Tainted: P           ---------------    2.6.32-431.23.3.1chaos.ch5.2.x86_64 #1
2015-04-01 20:44:00 Call Trace:
2015-04-01 20:44:00  [<ffffffffa049f0ad>] ? spl_dumpstack+0x3d/0x40 [spl]
2015-04-01 20:44:00  [<ffffffffa049f2a2>] ? spl_panic+0xc2/0xe0 [spl]
2015-04-01 20:44:00  [<ffffffff8152cf91>] ? mutex_lock+0x31/0x60
2015-04-01 20:44:00  [<ffffffff8152cf91>] ? mutex_lock+0x31/0x60
2015-04-01 20:44:00  [<ffffffffa0630b88>] ? dbuf_dest+0x48/0x70 [zfs]
2015-04-01 20:44:00  [<ffffffffa049be06>] ? spl_kmem_cache_free+0x46/0x210 [spl]
2015-04-01 20:44:00  [<ffffffffa062e00c>] ? dbuf_destroy+0x1dc/0x280 [zfs]
2015-04-01 20:44:00  [<ffffffffa062f18e>] ? dbuf_evict+0x1e/0x30 [zfs]
2015-04-01 20:44:00  [<ffffffffa0650469>] ? dnode_evict_bonus+0x119/0x180 [zfs]
2015-04-01 20:44:00  [<ffffffffa062f639>] ? dbuf_rele_and_unlock+0x3b9/0x4c0 [zfs]
2015-04-01 20:44:00  [<ffffffffa064cfa1>] ? dnode_setdirty+0x181/0x260 [zfs]
2015-04-01 20:44:00  [<ffffffffa064ffbb>] ? dnode_free_range+0x5db/0x6f0 [zfs]
2015-04-01 20:44:00  [<ffffffff8152cf91>] ? mutex_lock+0x31/0x60
2015-04-01 20:44:00  [<ffffffffa062fc16>] ? dbuf_rele+0x56/0x70 [zfs]
2015-04-01 20:44:00  [<ffffffffa062fc3e>] ? dmu_buf_rele+0xe/0x10 [zfs]
2015-04-01 20:44:00  [<ffffffffa06741fd>] ? sa_handle_destroy+0x7d/0x100 [zfs]
2015-04-01 20:44:00  [<ffffffffa06d03fd>] ? zfs_znode_dmu_fini+0x1d/0x30 [zfs]
2015-04-01 20:44:00  [<ffffffffa06d0ee8>] ? zfs_znode_delete+0xe8/0x220 [zfs]
2015-04-01 20:44:00  [<ffffffffa06b1b8a>] ? zfs_rmnode+0x22a/0x3a0 [zfs]
2015-04-01 20:44:00  [<ffffffffa06d0c84>] ? zfs_zinactive+0x134/0x2b0 [zfs]
2015-04-01 20:44:00  [<ffffffffa06cf99f>] ? zfs_inactive+0x7f/0x2a0 [zfs]
2015-04-01 20:44:00  [<ffffffffa06e6750>] ? zpl_inode_delete+0x0/0x30 [zfs]
2015-04-01 20:44:00  [<ffffffffa06e6676>] ? zpl_clear_inode+0x36/0x50 [zfs]
2015-04-01 20:44:00  [<ffffffff811a74e2>] ? clear_inode+0xd2/0x160
2015-04-01 20:44:00  [<ffffffffa06e6770>] ? zpl_inode_delete+0x20/0x30 [zfs]
2015-04-01 20:44:00  [<ffffffff811a7bce>] ? generic_delete_inode+0xde/0x1d0
2015-04-01 20:44:00  [<ffffffffa06e678e>] ? zpl_drop_inode+0xe/0x10 [zfs]
2015-04-01 20:44:00  [<ffffffff811a6b62>] ? iput+0x62/0x70
2015-04-01 20:44:00  [<ffffffff8119b809>] ? do_unlinkat+0x1a9/0x260
2015-04-01 20:44:00  [<ffffffff8100c755>] ? math_state_restore+0x45/0x60
2015-04-01 20:44:00  [<ffffffff8119baf2>] ? sys_unlinkat+0x22/0x40
2015-04-01 20:44:00  [<ffffffff8100b0b2>] ? system_call_fastpath+0x16/0x1b
```

**Comment by odoucet** [comment]
Hello,

I encountered a similar problem, and I'm not sure if this is the one or not. 
On SPL/ZFS 0.6.4.2-1 : 
```
VERIFY3(((*(volatile typeof((&db->db_mtx)->m_owner) *)&((&db->db_mtx)->m_owner))) == ((void *)0)) failed (ffff8818054b2740 ==           (null))
PANIC at dbuf.c:105:dbuf_dest()
Showing stack for process 39005
CPU: 1 PID: 39005 Comm: kworker/1:0 Tainted: P           O 3.10.76-1.el6.elrepo.x86_64 #1
Hardware name: Dell Inc. PowerEdge R420/0CN7CM, BIOS 2.1.3 05/21/2014
Workqueue: events delayed_fput
 ffffffffa039960c ffff8827d2d597d8 ffffffff815ecf8a ffff8827d2d597e8
 ffffffffa027c084 ffff8827d2d59988 ffffffffa027c2b2 0000000000000000
 ffffffffa03a5820 0000000000000030 ffff8827d2d59998 ffff8827d2d59938
Call Trace:
 [<ffffffff815ecf8a>] dump_stack+0x19/0x1f
 [<ffffffffa027c084>] spl_dumpstack+0x44/0x50 [spl]
 [<ffffffffa027c2b2>] spl_panic+0xc2/0x100 [spl]
 [<ffffffff81180294>] ? kmem_cache_free+0x114/0x400
 [<ffffffff815ed84e>] ? mutex_lock+0x1e/0x50
 [<ffffffffa02dfa18>] dbuf_dest+0x48/0x70 [zfs]
 [<ffffffffa0277c95>] spl_kmem_cache_free+0x45/0x170 [spl]
 [<ffffffffa02dd21b>] dbuf_destroy+0x1bb/0x1e0 [zfs]
 [<ffffffffa02de04e>] dbuf_evict+0x1e/0x30 [zfs]
 [<ffffffffa02fe17b>] dnode_evict_bonus+0x10b/0x150 [zfs]
 [<ffffffffa02de467>] dbuf_rele_and_unlock+0x357/0x3f0 [zfs]
 [<ffffffffa02fa740>] ? dnode_setdirty+0x170/0x1f0 [zfs]
 [<ffffffffa02fd2db>] ? dnode_free_range+0x57b/0x690 [zfs]
 [<ffffffffa02fc81e>] ? dnode_hold_impl+0x3ae/0x550 [zfs]
 [<ffffffff815ed84e>] ? mutex_lock+0x1e/0x50
 [<ffffffffa02de990>] dbuf_rele+0x40/0x50 [zfs]
 [<ffffffffa02de9ae>] dmu_buf_rele+0xe/0x10 [zfs]
 [<ffffffffa031f685>] sa_handle_destroy+0x75/0xc0 [zfs]
 [<ffffffffa0377c4d>] zfs_znode_dmu_fini+0x1d/0x30 [zfs]
 [<ffffffffa037875f>] zfs_znode_delete+0xcf/0x1b0 [zfs]
 [<ffffffffa0359301>] zfs_rmnode+0x211/0x340 [zfs]
 [<ffffffffa0378686>] zfs_zinactive+0x166/0x170 [zfs]
 [<ffffffffa037724c>] zfs_inactive+0x7c/0x250 [zfs]
 [<ffffffffa038cfce>] zpl_evict_inode+0x4e/0x70 [zfs]
 [<ffffffff811b5e64>] evict+0xb4/0x1d0
 [<ffffffff811b6063>] iput_final+0xe3/0x170
 [<ffffffff811b612e>] iput+0x3e/0x50
 [<ffffffff811b3b88>] dput+0x1d8/0x2c0
 [<ffffffff8119dd38>] __fput+0x188/0x270
 [<ffffffff8119df2a>] delayed_fput+0x9a/0xb0
 [<ffffffff8107b07d>] process_one_work+0x17d/0x500
 [<ffffffff8107c3bc>] worker_thread+0x21c/0x400
 [<ffffffff8107c1a0>] ? manage_workers+0x120/0x120
 [<ffffffff8108218e>] kthread+0xce/0xe0
 [<ffffffff810820c0>] ? kthread_freezable_should_stop+0x70/0x70
 [<ffffffff815f9508>] ret_from_fork+0x58/0x90
 [<ffffffff810820c0>] ? kthread_freezable_should_stop+0x70/0x70
```

When this happens, zpool status runs fine, zfs list runs fine, but nfsd processes are in D state : 
```
[<ffffffffa027e1a5>] cv_wait_common+0x115/0x130 [spl]
[<ffffffffa027e215>] __cv_wait+0x15/0x20 [spl]
[<ffffffffa033891b>] txg_wait_open+0x8b/0xd0 [zfs]
[<ffffffffa02f670f>] dmu_tx_wait+0x23f/0x250 [zfs]
[<ffffffffa02f6d6c>] dmu_tx_assign+0x3c/0x80 [zfs]
[<ffffffffa0375566>] zfs_write+0x436/0xb80 [zfs]
[<ffffffffa038bbbf>] zpl_iter_write_common+0xef/0x150 [zfs]
[<ffffffffa038bc32>] zpl_aio_write+0x12/0x20 [zfs]
[<ffffffff8119b4ad>] do_sync_readv_writev+0x6d/0xa0
[<ffffffff8119c7bb>] do_readv_writev+0xfb/0x2f0
[<ffffffff8119c9f6>] vfs_writev+0x46/0x60
[<ffffffffa080dfe1>] nfsd_vfs_write+0xa1/0x2a0 [nfsd]
[<ffffffffa080e8e7>] nfsd_write+0xe7/0x100 [nfsd]
[<ffffffffa08177ff>] nfsd3_proc_write+0x9f/0xd0 [nfsd]
[<ffffffffa0809258>] nfsd_dispatch+0xa8/0x1b0 [nfsd]
[<ffffffffa07a951c>] svc_process_common+0x30c/0x530 [sunrpc]
[<ffffffffa07a9a92>] svc_process+0x112/0x160 [sunrpc]
[<ffffffffa08099af>] nfsd+0xbf/0x130 [nfsd]
[<ffffffff8108218e>] kthread+0xce/0xe0
[<ffffffff815f9508>] ret_from_fork+0x58/0x90
[<ffffffffffffffff>] 0xffffffffffffffff
```

Is this also related to #3443 , and then fixed in 0.6.5 ?

**Comment by behlendorf** [comment]
Closing this was resolved.
